### DISCUSSION TITLE
What's the difference between prediction and test?

### DISCUSSION THREAD

**Discussion by yuvalsCG** [Lightning-AI/pytorch-lightning#11455] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
What's the difference between prediction and test:
predict_step vs test_step?
trainer.predict() vs trainer.test()?

In what use cases should I prefer using the one over the other?

**Comment by ananthsub** [comment]
If you mean when to use test vs predict, I've traditionally seen it this way:
Test is used for a holdout section of your dataset, used to evaluate a model after you've done hyperparameter tuning for fair comparison. The test set has labels included

Prediction: There are no labels available, only the model inputs, meaning prediction isn't used for evaluation

Regarding differences in Lightning, the two code paths are pretty similar are very similar. The model is put into eval mode, gradients are disabled, and the trainer makes one pass through the corresponding dataloader(s), calling the relevant hooks in the Lightning Module or callback (prefixed with test or predict)

**Reply by adosar** [reply]
Are you sure that model is in `eval` mode during `predict`? This is nowhere documented.
